File: projecteuler/Euler004.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
  public static boolean threeDigits(int n) {
    return n > 99 && n < 1000;
  }

  public static int palindromeNumber(int n) {
    return (n / 100) * 100000 +
      ((n % 100) / 10) * 10000 +
      (n % 10) * 1000 +
      (n % 10) * 100 +
      ((n % 100) / 10) * 10 +
      n / 100;
  }

  public static boolean threeDigitFactors(int n) {
    int i = 2;
    while (i * i <= n) {
      if (n % i == 0) {
        if (threeDigits(i) && threeDigits(n / i)) {
          return true;
        }
      }
      i += 1;
    }
    return false;
  }

  public static ArrayList<Integer> generatePalindromes() {
    ArrayList<Integer> palindromes = new ArrayList<Integer>();

    for (int i = 101; i < 1000; i++) {
      int palindrome = palindromeNumber(i);
      if (threeDigitFactors(palindrome)) {
        palindromes.add(palindrome);
      }
    }

    return palindromes;
  }

  public static int findPalindrome(int n, ArrayList<Integer> palindromes) {
    int palindrome = 0;
    int s = 0;
    int e = palindromes.size() - 1;
    int m;

    while (s <= e) {
      m = (s + e) >> 1;
      if (palindromes.get(m) < n) {
        palindrome = palindromes.get(m);
        s = m + 1;
      } else {
        e = m - 1;
      }
    }

    return palindrome;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int t = scanner.nextInt();
    int n;
    ArrayList<Integer> palindromes = generatePalindromes();

    for (int i = 0; i < t; i++) {
      n = scanner.nextInt();
      System.out.println(findPalindrome(n, palindromes));
    }
  }
}

File: projecteuler/euler004.rb
def self.palindrome_number(n)
  (n / 100) * 100000 +
  ((n % 100) / 10) * 10000 +
  (n % 10) * 1000 +
  (n % 10) * 100 +
  ((n % 100) / 10) * 10 +
  n / 100
end

def three_digits?(n)
  n > 99 && n < 1000
end

def self.three_digit_factors(n)
  factors_found = false

  i = 2
  while i * i <= n do
    if n % i == 0
      l, r = i, n / i
      if three_digits?(l) && three_digits?(r)
        factors_found = true
        break
      end
    end
    i += 1
  end

  factors_found
end

def self.generate_palindromes
  palindromes = []
  i = 101
  while i < 1000 do
    palindrome = palindrome_number(i)
    palindromes.push(palindrome) if three_digit_factors(palindrome)
    i += 1
  end

  palindromes
end

palindromes = generate_palindromes

gets.to_i.times do
  palindrome = -1
  n = gets.to_i

  s, e = 0, palindromes.length - 1
  while s <= e do
    m = (s + e) >> 1
    if palindromes[m] < n
      palindrome = palindromes[m]
      s = m + 1
    else
      e = m - 1
    end
  end

  puts palindrome
end
